Rewrite

lukaszx0 opened the pull request:
Hey,

I've been recently working on a new version (`v0.3.x`) which changes a lot of fundamental things. It's not yet finished however the MVP is working.

In new version, I plan to ditch rails engine-related stuff (it always seemed like a hack and was somewhat limiting) and make this gem a simple railtie with some custom additions and concepts like modules and dependency injection.

Check out updated w/ new section README and let me know what you think!

@xaviershay @zanker @ggilder @tarcieri @zachmargolis

lukaszx0 committed:
Add dependency graph class

We'll use it to resolve module dependencies

lukaszx0 committed:
Add container class for managing module lifecycle

lukaszx0 committed:
Test injecting deps

lukaszx0 committed:
Test injection better, rename to real-world names

lukaszx0 committed:
Create attr_accessor for each dependency when it's defined

lukaszx0 committed:
Introduce to_module

Instead of injecting Module class, we're calling #to_module on it, which should return
canonical object of module, instead of module wrapper with lifecycle methods

lukaszx0 committed:
Add sample modules

lukaszx0 committed:
Remove unused method

lukaszx0 committed:
Add todo for auto-created attrs

lukaszx0 committed:
Inject deps as #init arguments

lukaszx0 committed:
TODOs and clean up

lukaszx0 committed:
Make everything private besides #run!

lukaszx0 committed:
Take options hash as an arg

lukaszx0 committed:
Add railtie, disable engine

lukaszx0 committed:
Bump version to 0.3.0

lukaszx0 committed:
Include default modules

lukaszx0 committed:
Disable engine stuff

lukaszx0 committed:
Missing AS require

lukaszx0 committed:
Better error checking

lukaszx0 committed:
Update default modules & tests

lukaszx0 committed:
Update README with new stuff introduced in 0.3

zachmargolis commented on README.md:
```diff
@@ -40,6 +40,61 @@ Or install it yourself as:
+        def init(config, logger)
+          self.config = config
+          self.logging = logging
+          self.ping_server = PingServer.new(config.ping_server, logger)
```
hey I know this is just an example but I am super confused by this.

Is this a class method for configuring something? Or should this just be `::PingServer` ?

lukaszx0 replied:
Ugh, my fault. It should be rather called `PingServerModule`, the wrapper with lifecycle methods and dep injection stuff which takes care of booting, staring and stoping `PingServer`. Fixed now, sorry for confusion!

lukaszx0 committed:
Fix readme, rename module

lukaszx0 committed:
Remove namespaces, from example

lukaszx0 committed:
Pass #app instance to the container, fix enabling modules

lukaszx0 committed:
Remove Modules namespace, add naming convention

lukaszx0 committed:
Add module naming convention to #_name, remove #resolve_module_name

lukaszx0 committed:
Add standalone modules to resolved modules

lukaszx0 committed:
Port _status logic from engine to standalone sinatra app

lukaszx0 committed:
Port _admin logic from engine to standalone sinatra module

lukaszx0 committed:
Add :stats and :admin modules to default ones

lukaszx0 committed:
Move admin assets and set :public_path for sinatra app

lukaszx0 committed:
Remove old engine-based code and structure

lukaszx0 committed:
Don't hardcode base paths

lukaszx0 committed:
Mount one sinatra 'router' application, and use subclasses of base app as middlewares

lukaszx0 committed:
Port admin app from engine, add method for managing sidebar

lukaszx0 committed:
Hack around multiple lookup paths issue in sinatra

lukaszx0 committed:
Enchance #get - allow setting sidebar link

lukaszx0 committed:
Store enabled modules in config

lukaszx0 committed:
Clean up config module, remove unused things

lukaszx0 committed:
load_initializer loads rails' initializer for rails_service in the orderd we want

lukaszx0 committed:
Collect all submodules in #start

This way, modules dependent on AdminModule have a chance to load their files in #init

lukaszx0 committed:
Add registry class

lukaszx0 committed:
init => start, just like for admin module

To let other modules load their apps in #init

lukaszx0 committed:
registry

lukaszx0 committed:
Remove leftovers, make test run & green again

tarcieri commented on README.md:
```diff
@@ -40,6 +40,55 @@ Or install it yourself as:
 
     $ gem install rails-service
 
+## Modules
```
The word "Module" already has a pretty precise and specific meaning as a first-class feature of the Ruby language. I think if you picked a different word for this purpose it would be less confusing.

Some suggestions:
- `Initializer`
- `Plugin`
- `Component`
- `Dependency`

lukaszx0 replied:
Yeah good call and I agree (it's surprising to me that I haven't had any issues with using it in a first place). I'll change it to either `Plugin` on `Component` (leaning towards former)

jhump replied:
This is funny because we had such similar naming conversations when trying to name sort-of-similar concepts in our hack week project, java service container v2. I think we landed with `Component` (once we ditched dagger2 anyway, which already had its own concept of "component"). Although `Plugin` was also on our list in that discussion, too.

tarcieri commented on README.md:
```diff
@@ -40,6 +40,55 @@ Or install it yourself as:
+
+ * Config - loads application config used by other modules
+ * Logging - sets up and configures logging for service
+ * Admin - mounts Sinatra admin app for your service under `/_admin` [TODO]
```
Crazy idea: Rails engine for admin! :open_mouth: 

Edit: well, it's not that crazy. The Resque admin UI got way to unwieldy and I ended up rewriting it as a Rails Engine at LivingSocial:

https://github.com/resque/resque-web

Like night and day code-quality wise...

lukaszx0 replied:
If you look at the beginnings of the project I've started with rails-service as it being a rails engine however this was a bit limiting and felt very hacky.

Having just admin panel as an rails-engine is something to consider though we would need to figure out how to make it easily extensible so it's possible to add new actions (w/ their views) to the mix from app level or "decorator" gem level.
